Full stages support for github actions

jakemac53 opened the pull request:
Closes https://github.com/google/mono_repo.dart/issues/273

Adds a `stages` key to the `github` config which allows configuring stage ordering and conditions in the same way as the key under the `travis` config.

Respects stage ordering by adding `needs` on all jobs from the previous stage.

kevmoo reviewed (commented):
Seems okay.

I'd like to see a few actual tests.

Can `stages:` be empty? Only a list or a map?

does name have to map to a defined stage?
What's valid values for `if`?
Can other keys be provided?

jakemac53 commented on mono_repo/README.md:
```diff
@@ -114,6 +114,15 @@ github:
+  stages:
+    - name: cron
+      # Only run this stage for scheduled cron jobs
+      if: github.event_name == 'schedule'
```
This is only for github, each CI provider configures this separately because they have different syntax.

I could also show an example for travis in the readme, but I think it would be duplicative?

kevmoo replied:
so...uh...why is this setting not under the `github` key?

jakemac53 replied:
It is

kevmoo replied:
Oh. 🤦

jakemac53 commented on .github/workflows/dart.yml:
```diff
@@ -251,3 +268,11 @@ jobs:
             "${CHAT_WEBHOOK_URL}"
         env:
           CHAT_WEBHOOK_URL: "${{ secrets.CHAT_WEBHOOK_URL }}"
+    needs:
```
Ya I could have listed _all_ jobs here but you only have to list all the jobs from the previous stage. Those depend themselves on their previous stages, etc, so this is less verbose but has the same behavior.

> Do we need to create two copies of this? One for each "set" of jobs?

Not sure what you mean by that?

kevmoo replied:
If job 1 fails – I don't think this will run is my point.

See what I mean?

jakemac53 replied:
Right, that is what stages are supposed to be and it retains the original behavior from before.

kevmoo replied:
understood.

But the purpose of having that 'on_conclusion' thingy is to apply when anything fails.

...but that's now broken w/ 'stages'

jakemac53 replied:
Ohhhhh I see what you are wondering about now.

We will have to test what happens here, I will make an artifical failure on this pr and see if we get notified.

jakemac53 replied:
Tested and it does work as intended, setting `if: failure()` for your completion job enables it to run whenever any dependent task fails, even if other dependent tasks are skipped

jakemac53 commented on mono_repo/lib/src/ci_shared.dart:
```diff
@@ -180,5 +183,79 @@ void logPackages(Iterable<PackageConfig> configs) {
+/// The [conditionalStages] are CI specific, as they use CI specific expression
+/// syntax.
+List<String> calculateOrderedStages(
+    RootConfig rootConfig, Map<String, ConditionalStage> conditionalStages) {
```
@munificent I blame you for this lol

We need a lint or something you can enforce here instead of relying on reviews cause ill never do this by default lol.

kevmoo replied:
Agreed we need a lint!

jakemac53 commented on mono_repo/lib/src/commands/github/github_yaml.dart:
```diff
@@ -364,3 +394,14 @@ class _SelfValidateJob implements HasStageName {
   _SelfValidateJob(this.stageName);
 }
+
+class MapEntryWithStage<K, V> implements MapEntry<K, V> {
```
I prefer it with the type arguments myself - the type looks nicer when you see it in code (it isn't hiding the useful information).

kevmoo commented on mono_repo/lib/src/commands/github/github_yaml.dart:
```diff
@@ -364,3 +394,14 @@ class _SelfValidateJob implements HasStageName {
   _SelfValidateJob(this.stageName);
 }
+
+class _MapEntryWithStage<K, V> implements MapEntry<K, V> {
```
@jakemac53  Remember – you can drop the type params here – we know what they are!

jakemac53 replied:
See my other comment (kind of lost now) but I prefer to keep them because otherwise you can't see the K/V types where this type is used. There isn't any downside really afaik?

kevmoo replied:
🤷 – not a biggy, I guess.

It makes things cleaner, I think

jakemac53 commented:
Added unit tests for errors.

> What's valid values for `if`?

Whatever expression syntax your CI accepts, its just copy/pasted in. I added a test to confirm string values.

kevmoo commented on mono_repo/test/github_command_test.dart:
```diff
@@ -332,7 +419,7 @@ line 3, column 5 of mono_repo.yaml: Unsupported value for "workflows". Workflows
 }
 
-Future<void> _testBadConfig(
+Future<void> _testBadConfigWithYamlException(
```
👍

kevmoo commented on mono_repo/test/github_command_test.dart:
```diff
@@ -343,3 +430,14 @@ Future<void> _testBadConfig(
   );
 }
+
+Future<void> _testBadConfigWithUserException(
```
👍

kevmoo commented on mono_repo/test/github_command_test.dart:
```diff
@@ -343,3 +430,14 @@ Future<void> _testBadConfig(
+
+Future<void> _testBadConfigWithUserException(
+    Object monoRepoYaml, Object expectedMessage,
+    {Object expectedDetails, Object printMatcher}) async {
```
Trailing comma 😁

jakemac53 replied:
omg lol

kevmoo reviewed (approved):
Oh, changelog + version change is also good.

😄 

Don't forget to re-run the build
